docs/reference/object-model.md

```diff
@@ -0,0 +1,105 @@
+
+A _path_ is a readable string, typically decoded as UTF-8.  lakeFS maps paths to their objects
+according to specific rules.  Importantly (and unlike many object stores), lakeFS may map
+multiple paths to the same object on backing storage.  lakeFS paths use the `lakefs` protocol,
+described below.
+
+## lakeFS the version control system
+
+lakeFS borrows its concepts for version control from Git.
+
+### Repository
+
+A _repository_ is a collection of objects with common history tracking.  lakeFS manages
+versions of the repository, identified by their commits.  A _commit_ is a collection of object
+metadata and data, including especially all paths and the object contents and metadata at that
+commit.  Commits have their own _commit metadata_, which includes a textual comment and
+additional user metadata.
+
+### Commits
+
+Commits are organized into a _history_ using their parent commits.  Every repository has
+exactly one _initial commit_ with no parents.  <span style="font-size: smaller">Note that a
+Git repository may have multiple initial commits.</span> A commit with more than one parent is
+a _merge commit_.  <span style="font-size: smaller">Currently lakeFS only supports merge
+commits with two parents.</span>
+
+### Identifying commits
+
+A commit is identified by its _commit ID_, a digest of all contents of the commit.  Commit IDs
+are by nature long, so a unique prefix may be used to abbreviate them (but note that short
+prefixes can become non-unique as the repository grows, so prefer to avoid abbreviating when
+storing commits for the long term).  A commit may also be identified by using a tag or a
+branch.  A repository can be read from any commit.
+
+A _tag_ is an immutable pointer to a single commit.  Tags have readable names.  Because tags
+are commits, a repository can be read from any tag.
+
+A _branch_ is a mutable pointer to a commit and its staging area.  Repositories are readable
+from any branch, but they are also **writable** to a branch.  The _staging area_ associated
+with a branch is mutable storage where objects can be created, updated or deleted.  These
+objects are readable when reading from the branch.  To **create** a commit from a branch, all
+files from the staging area are merged into the contents of the current branch, creating the
+new set of objects.  The parent of the commit is the previous branch tip, and the new branch
+tip is set to this new commit.
+
+The _history_ of the branch is the list of commits from the branch tip through the first
+parent of each commit.  Histories go back in time.
+
+The other way to create a commit is to merge an existing commit onto a branch.  To _merge_ a
+source commit into a branch, lakeFS finds the closest common ancestor of that source commit
+and the branch tip, called the "base".  Then it performs a 3-way merge.
+
+## Concepts unique to lakeFS
+
+_Underlying storage_ is the area on some other object store that lakeFS uses to store object
```

Comment:
I think that underlying storage is the right term to use here but as a user, I may be confused with other terms used by our system. Such as storage namespace (in the UI, which is the underlying bucket on the underlying storage), and blockstore (used in the configurations file to specify the underlying storage).

Susggestions: 
- Add the definition of storage namespace to this document, and consider renaming it in the UI to "Underlying bucket"
- Consider changing the blockstore key in the config file to underlyingstorage, or add a reference to the blockstore term in this doc to avoid the breaking change.

Reply:
Not done.  This is a product change that may be warranted.  I would like to handle it as such, and particularly *not* make the data model depend on this product change.  I _will_ change this documentation when and if we change this naming.  Note in particular that changing the name of a configuration section requires much attention to backwards compatibility, and I don't want to delay this document for that.

Reply:
makes sense to most of the items I mentioned, but isn't a storage namespace (in the context we use it) a concept that is special to lakefs?